### DISCUSSION TITLE
Laravel initialization with rsync

### DISCUSSION THREAD

**Discussion by bilogic** [deployphp/deployer#2565] [Help needed] [answered]
Hi,

1. My custom script uses rsync to deploy code to production which uses private packages
2. So, I'm hoping I can still deploy purely using rsync (and not rely on composer to pull in the private packages etc)
3. My issue is, for a brand new server, deployer throws errors after the first rsync as files like `.env` are still not provided yet
4. Is there some way to configure 2 different `add('rsync', [...`?
5. What I'm hoping to do is `rsync1`, then I manually log into the server to provide the `.env`, then perform `rsync2`, make sense?

```PHP
task('init', [
    ...
    'rsync1', // rsync folders (excluding vendor), artisan config:clear, artisan key:generate
    ...
]);

task('deploy', [
    ...
    'rsync2', // rsync folders (including vendor)
    ...
]);
```

**Comment by antonmedv** [comment]
This is totally doable. Just write custom task.

**Reply by bilogic** [reply]
Thanks!

**Comment by bilogic** [comment]
Ok, I managed to figure out
```PHP
task('config_rsync1', function() {
    add('rsync', [...]);
});

task('init', [
    ...
    'config_rsync1',
    'rsync', // rsync folders (excluding vendor), artisan config:clear, artisan key:generate
    ...
    'config_rsync2',
    'rsync',
]);

```

**Reply by bilogic** [reply]
@antonmedv 

What I did here did not seem to work. If I exclude `vendor` in `config_rsync1`, it is still excluded in `config_rsync2`. Is there a way for `config_rsync2` reset the changes in `config_rsync1`?

Thank you.

**Reply by antonmedv** [reply]
It’s possible vis on() function. But looks like you’re doing something wrong.

**Reply by bilogic** [reply]
Possible to illustrate the code structure? 

By the way, both `config_rsync1` and `config_rsync2` rsync with the same remote host.
I had to break them into 2 because some folders required `--copy-links` while folders like `vendor` did not.

Thank you.

**Reply by bilogic** [reply]
Looking at https://github.com/deployphp/deployer/blob/master/contrib/rsync.php, it does not seem possible to rsync to the same remote multiple times with different configuration.

Am I wrong?

**Reply by bilogic** [reply]
I tried `upload()`, but `options` does not seem to be passed in as well, see https://github.com/deployphp/deployer/issues/2568

```
task('rsync:vendor', function () {
    upload('vendor', '{{ release_path }}', [
        'options' => [
            'update',
            // 'times',
            'no-p',
            'progress',
            'delete',
            'copy-links',
        ]
    ]);
});
```

**Comment by bilogic** [comment, marked as answer]
Use multiple `upload()` (which uses `Rsync.php` instead of `contrib/rsync.php`)

Also take note that `upload()` requires the use of `--` for it's options unlike `contrib/rsync.php`

e.g. `$config = ['options' => ['--copy-links']];`

> https://deployer.org/docs/api.html#upload
